File: Java/labs/lab5/task_2_2 – solution/src/BookCollection.java
import java.util.ArrayList;

public class BookCollection implements Readable {

  private final ArrayList<Readable> children = new ArrayList<>();

  public ArrayList<Readable> getChildren() {
    return children;
  }

  public void add(Readable readable) {
    children.add(readable);
  }

  public void remove(Readable readable){
    children.remove(readable);
  }


  @Override
  public void checkout() {
    for (Readable child : children) {
      child.checkout();
    }
  }

  @Override
  public void returnBook() {
    for (Readable child : children) {
      child.returnBook();
    }

  }
}

File: Cpp/lab09_Composite/task_2_2 - solution/src/book_collection.cpp
#include <algorithm>

#include "book_collection.hpp"


BookCollection::BookCollection() {}

std::vector<Readable*> const& BookCollection::getChildren() const
{
  return this->children;
}

void BookCollection::add(Readable* readable)
{
  this->children.push_back(readable);
}

void BookCollection::remove(Readable* readable)
{
  std::remove(
    this->children.begin(),
    this->children.end(),
    readable
  );
}

void BookCollection::checkout()
{
  for (auto & child: this->children)
  {
    child->checkout();
  }
}

void BookCollection::returnBook()
{
  for (auto & child: this->children)
  {
    child->returnBook();
  }
}
